### PR TITLE
Add best-effort coverage reports per test type

### PR DESCRIPTION
**Pull request by jsoriano**
Ensure that each executed test type marks some content as covered, so we can have an idea of the files covered by tests, and the test types executed.

Currently we report if a test type has been executed by marking a line in a manifest as covered, but this fails if the manifest doesn't have enough lines. This change attempts to solve this kind of issues.

At the moment the following files are fully or partially marked as covered if an specific type of test is executed:
* Pipeline tests:
  - [x] No changes in this PR, we have coverage based on ES stats.
* Asset tests:
  - [x] All files under the kibana directory.
* Policy tests:
  - [x] All manifests.
  - [x] All template files in the executed data stream.
* Static tests
  - [x] Sample events.
  - [x] ~~Benchmark configs.~~ No need to cover dev files.
* System tests
  - [x] All manifests.
  - [x] Fields files.

The rest of non-development files are marked as not covered.

Fixes https://github.com/elastic/elastic-package/issues/1624.

### CODE REVIEW

**Comment by jsoriano** [comment]
test integrations

**Comment by jsoriano** [comment]
test integrations

**Comment by jsoriano** [comment]
test integrations

**Comment by jsoriano** [comment]
test integrations

File: cmd/check.go [hunk 1]
```diff
@@ -24,7 +24,6 @@ func setupCheckCommand() *cobraext.Command {
 		Args:  cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
 			err := cobraext.ComposeCommands(args,
-				setupFormatCommand(),
```
Comment (jsoriano) [inline cmd/check.go:27]: We can make this change in a different PR: https://github.com/elastic/elastic-package/pull/1936

**Comment by jsoriano** [comment]
Ok, it is working now. I will open for review. There is a working report in https://sonar.elastic.dev/component_measures?metric=new_conditions_to_cover&view=list&pullRequest=10273&id=elastic_integrations_AYu5LCaKQZlFqhqWIrk_

Perhaps we can review the files added to the coverage and we will need to review the list of files ignored by sonarqube in the integrations repository, the latter will be done in a different PR.

File: internal/testrunner/coveragereport.go [hunk 2]
```diff
@@ -33,20 +31,6 @@ func CoverageFormatsList() []string {
...
-		"system":   3,
-		"static":   4,
-	}
-	lineNumber, ok := lineNumberPerTestType[testType]
-	if !ok {
-		lineNumber = 5
-	}
-	return lineNumber
-}
-
```
Comment (mrodm) [inline internal/testrunner/coveragereport.go:49]: :fire:

File: internal/testrunner/coveragereport_test.go [hunk 3]
```diff
@@ -212,49 +79,14 @@ func TestCreateCoverageReport(t *testing.T) {
...
-				TestType:  "Coverage for asset test",
-				Timestamp: 10,
+				TestType: "Coverage for system test",
 			},
 		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			report, err := createCoverageReport(tt.rootPath, tt.packageName, tt.packageType, tt.testType, tt.results, tt.coverageFormat, tt.timestamp)
```
Comment (mrodm) [inline internal/testrunner/coveragereport_test.go:257]: Is timestamp not mandatory for the coverage reports using Cobertura?

File: internal/testrunner/coverage.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/elastic/elastic-package/internal/files"
+)
+
+// GenerateBasePackageCoverageReport generates a coverage report where all files under the root path are
+// marked as not covered. It ignores files under _dev directories.
+func GenerateBasePackageCoverageReport(pkgName, rootPath, format string) (CoverageReport, error) {
```
Comment (mrodm) [inline internal/testrunner/coverage.go:23]: Could this function solved also this issue for pipeline tests ? 
https://github.com/elastic/elastic-package/issues/1624

If all files present in the package are marked as uncovered, there will be coverage (zero) for those files without tests.

Just one concern regarding also about pipelines, if there is a coverage report created by pipeline in some package. With this change there would be lines marked as uncovered (that previously there was no information), that could cause a decrease in the coverage (%). Would that be an issue?

File: internal/testrunner/coveragereport_test.go (hunk 3 above)
Comment (jsoriano) [reply internal/testrunner/coveragereport_test.go:257]: Maybe, but there is no coverage report generated by this method now, it uses the coverage in the results, and then the timestamp is not used anymore, this is why I removed the parameter.

File: internal/testrunner/testrunner.go [hunk 5]
```diff
@@ -145,6 +145,23 @@ func NewResultComposer(tr TestResult) *ResultComposer {
...
+
+// CoveragePackageName returns a package name that can be used in coverage reports, based on information
+// in the composer.
+func (rc *ResultComposer) CoveragePackageName() string {
+	if rc.DataStream != "" {
+		return rc.Package + "." + rc.DataStream
+	}
+
+	return rc.Package
+}
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply internal/testrunner/testrunner.go:163]: Done in 51a85a356b7804f41ebdd6168e9f4422c14c21be.

File: internal/testrunner/coverage.go (hunk 4 above)
Comment (jsoriano) [reply internal/testrunner/coverage.go:23]: > Could this function solved also this issue for pipeline tests ?
> #1624

Probably yes. Tests will generate coverage for all files now, even if it is empty.

> If all files present in the package are marked as uncovered, there will be coverage (zero) for those files without tests.

Correct, if there are no tests, there is no coverage.

> Just one concern regarding also about pipelines, if there is a coverage report created by pipeline in some package. With this change there would be lines marked as uncovered (that previously there was no information), that could cause a decrease in the coverage (%). Would that be an issue?

Yes, there can be a decrease in coverage in several cases. I wouldn't consider this as an issue. Before there was more coverage % just because we were not reporting coverage for many files. We will probably revisit what files we mark as covered or uncovered, and how.

File: internal/testrunner/runners/pipeline/tester.go [hunk 6]
```diff
@@ -300,74 +300,67 @@ func (r *tester) checkElasticsearchLogs(ctx context.Context, startTesting time.T
...
-	if e, ok := err.(testrunner.ErrTestCaseFailed); ok {
-		tr.FailureMsg = e.Error()
-		tr.FailureDetails = e.Details
-		return tr, nil
-	}
 	if err != nil {
-		err := fmt.Errorf("verifying test result failed: %w", err)
-		tr.ErrorMsg = err.Error()
-		return tr, nil
+		results, _ := rc.WithErrorf("verifying test result failed: %w", err)
```
[the remark this answers is not in the document]
Comment (jsoriano) [reply internal/testrunner/runners/pipeline/tester.go:346]: This is interesting, because what we are doing now is quite similar, `WithErrorf` includes a similar logic, that should be setting the failure fields:
```
        var tcf *ErrTestCaseFailed
        if errors.As(err, &tcf) {
                rc.FailureMsg += tcf.Reason
                rc.FailureDetails += tcf.Details
                return []TestResult{rc.TestResult}, nil
        }
```
Taking a look...

[the remark this answers is not in the document]
Comment (jsoriano) [reply internal/testrunner/runners/pipeline/tester.go:346]: Condition was incorrect, the variable doesn't have to be a pointer. Apart of that I adjusted when the prefix `test case failed:` is expected.
